### DISCUSSION TITLE
Page Layout: incorrect layouting behavior?

### DISCUSSION THREAD

**Discussion by hroemer** [zauberzeug/nicegui#216] [Q&A]
Experimenting with the example shown in #187 I fear the generated layout is not correct when defining `top_corner` or `bottom_corner` as `False`:

```python
from fastapi import FastAPI
from nicegui import ui

def create_layout(app: FastAPI, path: str = 'clean', title: str = 'clean'):
    @ui.page(path=path, dark=True, title=title)
    async def page_layout():
        ui.label('CONTENT')
        with ui.header():
            ui.label('HEADER')
        with ui.left_drawer(top_corner=True, bottom_corner=True):
            ui.label('LEFT DRAWER')
        with ui.right_drawer(top_corner=False, bottom_corner=False):
            ui.label('RIGHT DRAWER')
        with ui.footer():
            ui.label('FOOTER')

    ui.run_with(app=app)


app = FastAPI()
layout = create_layout(app=app, path='/layout', title='layout')
```

The resulting HTML is:

```html
<div class="q-layout q-layout--standard" tabindex="-1" id="0" style="min-height: 711.333px;">
    <!----><!---->
    <div class="q-page-container" id="1" style="padding-top: 53px; padding-bottom: 53px;">
        <div id="2" class="q-pa-md column items-start gap-4">
            <div id="3">CONTENT</div>
            <header class="q-header q-layout__section--marginal fixed-top q-pa-md row items-start gap-4" id="4">
                <div id="5">HEADER</div>
                <!---->
            </header>
            <div class="q-drawer-container">
                <div class="q-drawer__opener fixed-left" aria-hidden="true"></div>
                <div class="fullscreen q-drawer__backdrop hidden" aria-hidden="true" style="background-color: rgba(0, 0, 0, 0);"></div>
                <aside class="q-drawer q-drawer--left q-drawer--dark q-dark q-layout--prevent-focus fixed q-drawer--on-top q-drawer--mobile q-drawer--top-padding" style="width: 300px; transform: translateX(-300px);">
                    <div id="6" class="q-drawer__content fit scroll q-pa-md"><div id="7">LEFT DRAWER</div></div>
                </aside>
            </div>
            <div class="q-drawer-container">
                <div class="q-drawer__opener fixed-right" aria-hidden="true"></div>
                <div class="fullscreen q-drawer__backdrop hidden" aria-hidden="true" style="background-color: rgba(0, 0, 0, 0);"></div>
                <aside class="q-drawer q-drawer--right q-drawer--dark q-dark q-layout--prevent-focus fixed q-drawer--on-top q-drawer--mobile q-drawer--top-padding" style="width: 300px; transform: translateX(300px);">
                    <div id="8" class="q-drawer__content fit scroll q-pa-md"><div id="9">RIGHT DRAWER</div></div>
                </aside>
            </div>
            <footer class="q-footer q-layout__section--marginal fixed-bottom q-pa-md row items-start gap-4" id="10">
                <!---->
                <div id="11">FOOTER</div>
            </footer>
        </div>
    </div>
</div>
```
The right drawer is drawn from top to bottom overlapping header and footer.

The layouting actually does not adhere to the [Quasar Layout component](https://quasar.dev/layout/layout) documentation.

Header, Footer, Drawer and PageContainer should be nested inside `<q-layout>` on the same level.
```html
<div class="q-pa-md">
   <q-layout view="lhh LpR lff" container style="height: 500px" class="shadow-2 rounded-borders">
      <q-header>...</q-header>
      <q-footer>...</q-footer>
      <q-page-container>
          <q-page>ACTUAL CONTENT HERE</q-page>
      </q-page-container>
     </q-layout>
</div>
```

1. Is this deviation from the Quasar an intended approach?
2. Would [nested-routes](https://quasar.dev/layout/routing-with-layouts-and-pages#nested-routes) and thus [SPA](https://developer.mozilla.org/en-US/docs/Glossary/SPA) still be an option here by defining the "outer layout" and just load individual views/pages inside the `<q-page-container>` / `<router-view>`?

**Comment by falkoschindler** [comment]
Strange. On my machine (MacOS with Chrome, Firefox or Safari) the layout looks as expected. The right drawer ends at the header and footer, respectively.

<img width="1171" alt="Screenshot 2023-01-03 at 18 34 43" src="https://user-images.githubusercontent.com/5767091/210410919-23cefb6a-9991-469e-a50b-af06c69fa92f.png">

Regarding question 1:
So you are suggesting to wrap the content in q-page-container and q-page, right? Honestly, I haven't thought about it, yet. I probably tried to place the content without unnecessary layout wrappers to reduce impact for use cases without any page layout in mind. But since we're using q-layout anyway, we can as well also use q-page-container and q-page. I'll try it out.

Regarding question 2:
Yes, we have SPAs in mind, but haven't yet worked on a concrete implementation. We definitely want to support it the not too distant future.

**Reply by falkoschindler** [reply]
Oh, now I see it: Currently header, footer and drawers are nested _inside_ q-page-container, which instead should be a sibling. I'll check if we can change the structure towards the recommendation.

**Comment by hroemer** [comment]
Thanks for investigating. It also works on my end if the browser can use the the full width:

![127 0 0 1_43434_layout_full](https://user-images.githubusercontent.com/16097755/210519373-99e1d8f9-4cc8-4254-ba6c-ee5565c3682a.png)

I am on Linux using a tiling window manager. I get the faulty behavior when not using full width:

![127 0 0 1_43434_layout_tiled](https://user-images.githubusercontent.com/16097755/210519357-10c92361-1004-43a9-a84e-2cb499140470.png)

It is also reproducible when scaling (e.g. 150% zoom factor in Chromium, 200% in Firefox). Thus I believe the CSS breakpoint (at 1024px) also has an impact.

Regarding question 1: if q-layout is already set, I would appreciate using quasar layouting recommendations. :+1: 

Regarding question 2: I haven't actually used [nested-routes](https://quasar.dev/layout/routing-with-layouts-and-pages#nested-routes) myself yet, but think this would help implementing seamless transitions between views.

**Comment by falkoschindler** [comment]
Very strange. I can't reproduce it. But maybe it helped to fix the layout hierarchy in commit https://github.com/zauberzeug/nicegui/commit/3eb10a9b1c3f2ab8787457e78304741439a6c0da.
@hroemer Do you still get the wrong layout with version 1.1.0?

**Reply by hroemer** [reply]
It still doesn't work as expected in version `1.1.2`:
![127 0 0 1_43434_layout](https://user-images.githubusercontent.com/16097755/212348907-50c4e3fa-5bca-4305-86eb-85bbe80d8499.png)

However, the rendered HTML changed and elements are now siblings:

```html
<div id="app" data-v-app="">
    <div class="q-layout q-layout--standard" >
        <div class="q-page-container">
            <main >
...
            </main>
        </div>
        <header>
...
       </header>
        <div class="q-drawer-container">
            <aside>
...
            </aside>
        </div>
        <div class="q-drawer-container">
            ...
        </div>
        <footer>
...
        </footer>
    </div>
</div>
``` 

Could the order (header, footer, drawer before page-container) be relevant?

**Reply by falkoschindler** [reply]
> Could the order (header, footer, drawer before page-container) be relevant?

To answer this question we could create a minimum example using plain HTML, JavaScript and Quasar. Then we can easily play with the order. If we only could replicate the problem, we would at least know where to search for a solution...

**Reply by hroemer** [reply]
Alright, I have a static example ready but currently trouble uploading it. Will get back to you, soon.

**Reply by hroemer** [reply]
I couldn't upload the files in the discussion thread, thus I added the static repro via gh pages.
You can access them now [here](https://hroemer.github.io/nicegui/quasar_layout/layout.html), the sources are located in the `doc` folder of my [fork](https://github.com/hroemer/nicegui/tree/quasar_layout/docs/quasar_layout).
I hope this helps.

**Reply by falkoschindler** [reply]
Does the [static demo](https://hroemer.github.io/nicegui/quasar_layout/layout.html) work or is the layout broken? I guess your main problem is that the right drawer does not disappear like the left one. But on my machine both drawers disappear at the same time.

Maybe you can try further reducing the code until the problem is gone. Or maybe the problem persists even after removing everything NiceGUI-related. Then it might be a Quasar issue.

**Reply by hroemer** [reply]
The layout is broken in the static demo as well. The right drawer should be limited by the header and the footer. It works for large page widths > 1024px. For lower widths the right drawer overlaps the header and the footer.

I further reduced the code. It contains only quasar and vue now. I even updated the quasar lib to the latest version and tried reordering the elements, but still the problem persists.

* [layout.html](https://hroemer.github.io/nicegui/quasar_layout/layout.html)
* [layout_order.html](https://hroemer.github.io/nicegui/quasar_layout/layout_order.html)

Could you please confirm that this is an issue on the quasar side? Maybe I totally missed something here!?

**Reply by falkoschindler** [reply]
I'm not sure. There's still NiceGUI's `renderRecursively` which _might_ do something wrong.
But maybe you can check from the other side: How does the [layout builder page](https://quasar.dev/layout-builder) behave? Do the drawers disappear correctly? Maybe you can export a layout and insert it into a plain JavaScript Vue app.
I'd love to debug this issue myself, but without being able to reproduce the problem I'm pretty blind.

**Reply by hroemer** [reply]
Thanks for pointing me to the layout builder. It actually shows the same behavior as with NiceGUI: https://user-images.githubusercontent.com/16097755/213912535-e2c469be-1446-4a91-a68a-9416e8806c1b.mp4

Now I get the impression that this is the desired behavior, which actually makes sense.
`hRf` for the right drawer would reduce the available space for `page` to an unusable size if the drawer insisted on not sliding over the page content for breakpoints < 1024px.

My apologies for causing noise and counterproductive efforts here! 🙇

**Reply by falkoschindler** [reply]
Oh I see! I wasn't aware that your screenshot showed the state _after pulling out the drawer_. I only shrunk my browser window and the drawers were gone as expected.
Anyway, glad to hear that everything is working now. 🙂
And thanks for _your_ effort! It could as well have been a subtle bug in NiceGUI only showing up in rare circumstances.
